server/internal/oauth/providers/gram.go

```diff
@@ -0,0 +1,112 @@
+type GramProvider struct {
+	logger   *slog.Logger
+	sessions *sessions.Manager
+}
+
+// NewGramProvider creates a new Gram OAuth provider
+func NewGramProvider(logger *slog.Logger, sessions *sessions.Manager) *GramProvider {
+	return &GramProvider{
+		logger:   logger,
+		sessions: sessions,
+	}
+}
+
+// BuildAuthorizationURL builds the authorization URL for Gram OAuth
+func (p *GramProvider) BuildAuthorizationURL(ctx context.Context, params AuthURLParams) (*url.URL, error) {
+	urlParams := url.Values{}
+	urlParams.Add("return_url", params.CallbackURL)
+	urlParams.Add("state", params.State)
+
+	// !TODO: Check why these are empty
+	// Set scope from provider configuration or request
+	// if len(params.ScopesSupported) > 0 {
+	// Gram provider doesn't use scopes in the same way, but we'll include it if specified
+	// } else if params.Scope != "" {
+	// Include scope if provided in request
+	// }
+
+	gramAuthURL := fmt.Sprintf("%s/v1/speakeasy_provider/login?%s",
+		params.SpeakeasyServerAddr,
+		urlParams.Encode())
+
+	authURL, err := url.Parse(gramAuthURL)
+	if err != nil {
+		p.logger.ErrorContext(ctx, "failed to parse gram OAuth URL", attr.SlogError(err))
+		return nil, fmt.Errorf("failed to parse gram OAuth URL: %w", err)
+	}
+
+	return authURL, nil
+}
+
+// ExchangeToken exchanges an authorization code for an access token from Gram
+func (p *GramProvider) ExchangeToken(
+	ctx context.Context,
+	code string,
+	provider repo.OauthProxyProvider,
+	toolset *toolsets_repo.Toolset,
+	serverURL *url.URL,
+) (*TokenExchangeResult, error) {
+	// Exchange code for ID token from Speakeasy
+	idToken, err := p.sessions.ExchangeTokenFromSpeakeasy(ctx, code)
+	if err != nil {
+		p.logger.ErrorContext(ctx, "failed to exchange code for token from oauth gram provider",
+			attr.SlogOAuthProvider(provider.Slug),
+			attr.SlogError(err))
+		return nil, fmt.Errorf("failed to exchange code for token: %w", err)
+	}
+
+	// Get user info from Speakeasy
+	userInfo, err := p.sessions.GetUserInfoFromSpeakeasy(ctx, idToken)
+	if err != nil {
+		p.logger.ErrorContext(ctx, "failed to get user info from oauth gram provider",
+			attr.SlogOAuthProvider(provider.Slug),
+			attr.SlogError(err))
+		return nil, fmt.Errorf("failed to retrieve user info: %w", err)
+	}
+
+	// Check if user has access to the organization
+	hasOrgAccess := false
+	for _, org := range userInfo.Organizations {
+		if org.ID == toolset.OrganizationID {
+			hasOrgAccess = true
+			break
+		}
+	}
+
+	if !hasOrgAccess {
+		p.logger.WarnContext(ctx, "user does not have access to organization",
+			attr.SlogOAuthProvider(provider.Slug),
+			attr.SlogUserID(userInfo.UserID),
+			attr.SlogOrganizationID(toolset.OrganizationID))
+		return nil, fmt.Errorf("user does not have access to the requested organization")
+	}
+
+	// Use idToken as access token for gram providers
+	return &TokenExchangeResult{
+		AccessToken: idToken,
+		ExpiresAt:   nil,
+	}, nil
```

Comment:
You could consider issuing an actual session here with the ID token. So that when you later get this token back and check whether the user has access to org. You are using the session and get to take advantage of user cache, etc.


	session := sessions.Session{
		SessionID:            idToken,
		UserID:               userInfo.UserID,
		ActiveOrganizationID: "",
	}
	
	s.sessions.StoreSession(ctx, session)